src/cache.c

```diff
@@ -0,0 +1,388 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#include <assert.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "cache.h"
+
+#define MIN(a, b) ((a < b) ? a : b)
+#define GOLDEN_RATIO_32 0x61C88647
+#define HASH(val) \
+    (((val) * (GOLDEN_RATIO_32)) >> (32 - (cache_size_bits))) % (cache_size)
```

Comment:
We can specify the restriction for `cache_size`, which MUST be a power of 2. Then, `% (cache_size)` can be replaced with `& (cache_size - 1)`.